### PR TITLE
Refactoring themes to be the sole method of changing formatting settings

### LINKED ISSUE #180: Add Apple News themes
Add the ability to create a theme as a set of Apple News plugin settings. This includes both the current settings and the JSON snippets that will be added as part of #193

### PR DESCRIPTION
**Pull request by bcampeau**
Closes #180

### CODE REVIEW

File: admin/partials/page_options_section.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
+<h3><?php echo esc_html( $section->name() ); ?></h3>
+<?php echo wp_kses_post( $section->get_section_info() ); ?>
+<table class="form-table apple-news">
+	<?php foreach ( $section->groups() as $group ): ?>
```
Comment (tomharrigan) [inline admin/partials/page_options_section.php]: space before `:`

File: admin/partials/page_options_section.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	<?php do_action( 'apple_news_before_setting_group', $group, false ); ?>
+	<tr>
+		<th scope="row"><?php echo esc_html( $group['label'] ); ?></th>
+		<td>
+			<fieldset>
+				<?php foreach ( $group['settings'] as $setting_name => $setting_meta ): ?>
```
Comment (tomharrigan) [inline admin/partials/page_options_section.php]: space before `:`

File: admin/partials/page_options_section.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+				<?php do_action( 'apple_news_before_setting', $setting_name, $setting_meta ); ?>
+				<label class="setting-container">
+					<?php if ( ! empty( $setting_meta['label'] ) ): ?>
```
Comment (tomharrigan) [inline admin/partials/page_options_section.php]: space before `:`

File: admin/partials/page_options_section.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+						<span class="label-name"><?php echo esc_html( $setting_meta['label'] ); ?></span>
+					<?php endif; ?>
+					<?php
+						echo wp_kses(
+							$section->render_field(
+								array(
+									$setting_name,
+									$setting_meta['default'],
+									$setting_meta['callback']
```
Comment (tomharrigan) [inline admin/partials/page_options_section.php]: trailing comma

File: admin/partials/page_options_section.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+							),
+							Admin_Apple_Settings_Section::$allowed_html
+						);
+					?>
+				</label>
+				<?php do_action( 'apple_news_after_setting', $setting_name, $setting_meta ); ?>
+				<br />
+				<?php endforeach; ?>
+
+				<?php if ( $group['description'] ): ?>
```
Comment (tomharrigan) [inline admin/partials/page_options_section.php]: space before `:`

File: admin/partials/page_theme_edit.php [hunk 6]
```diff
@@ -0,0 +1,20 @@
+<div class="wrap apple-news-theme-edit">
+	<form method="post" action="" id="apple-news-theme-edit-form">
+		<?php wp_nonce_field( 'apple_news_save_edit_theme' ); ?>
+		<input type="hidden" name="action" value="apple_news_save_edit_theme" />
+		<p>
+			<label id="apple_news_theme_name_label" for="apple_news_theme_name"><?php echo esc_html_e( 'Theme Name', 'apple-news' ) ?></label>
```
Comment (tomharrigan) [inline admin/partials/page_theme_edit.php]: redundant. `echo esc_html_e`

File: admin/partials/page_themes.php [hunk 7]
```diff
@@ -1,3 +1,4 @@
+<?php $themes = new \Admin_Apple_Themes(); ?>
```
Comment (tomharrigan) [inline admin/partials/page_themes.php]: worth checking that `$themes` contains a valid object before continuing?

File: admin/settings/class-admin-apple-settings-section-formatting.php [hunk 8]
```diff
@@ -619,25 +630,37 @@ public function after_section() {
 	 * @static
 	 * @access public
```
Comment (tomharrigan) [inline admin/settings/class-admin-apple-settings-section-formatting.php]: @param boolean $hidden

File: admin/settings/class-admin-apple-settings-section.php [hunk 9]
```diff
@@ -418,14 +452,21 @@ class Admin_Apple_Settings_Section extends Apple_News {
...
 	 */
-	function __construct( $page ) {
-		$this->page             = $page;
-		$base_settings          = new \Apple_Exporter\Settings;
-		self::$base_settings    = $base_settings->all();
-		$this->settings         = apply_filters( 'apple_news_section_settings', $this->settings, $page );
-		$this->groups           = apply_filters( 'apple_news_section_groups', $this->groups, $page );
-		self::$fonts            = apply_filters( 'apple_news_fonts_list', self::$fonts );
+	function __construct( $page, $hidden = false, $save_action = 'apple_news_options', $section_option_name = null ) {
+		$this->page             		= $page;
```
Comment (tomharrigan) [inline admin/settings/class-admin-apple-settings-section.php]: tab spacing funky in this block

File: tests/admin/test-class-admin-apple-themes.php [hunk 10]
```diff
@@ -58,6 +69,8 @@ public function testCreateTheme() {
 		$themes = new \Admin_Apple_Themes();
 		$current_settings = $this->settings->all();
 
+//		print_r( $themes->get_theme( $name ) );
```
Comment (tomharrigan) [inline tests/admin/test-class-admin-apple-themes.php]: delete

File: admin/partials/page_themes.php (hunk 7 above)
Comment (bcampeau) [inline admin/partials/page_themes.php]: No it would be impossible.

File: admin/settings/class-admin-apple-settings-section.php [hunk 11]
```diff
@@ -418,14 +452,21 @@ class Admin_Apple_Settings_Section extends Apple_News {
...
+	 * @param string $section_option_name
 	 */
-	function __construct( $page ) {
-		$this->page             = $page;
-		$base_settings          = new \Apple_Exporter\Settings;
-		self::$base_settings    = $base_settings->all();
-		$this->settings         = apply_filters( 'apple_news_section_settings', $this->settings, $page );
-		$this->groups           = apply_filters( 'apple_news_section_groups', $this->groups, $page );
-		self::$fonts            = apply_filters( 'apple_news_fonts_list', self::$fonts );
+	function __construct( $page, $hidden = false, $save_action = 'apple_news_options', $section_option_name = null ) {
```
Comment (tomharrigan) [inline admin/settings/class-admin-apple-settings-section.php]: spacing still funky for this and the next 10 lines

Comment (bcampeau) [inline admin/settings/class-admin-apple-settings-section.php]: looks fine in editor, i think it's just github

**Pull request merged**